File: mcp/src/tools/peers.ts
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { ToolContext } from "../types.js";
import { textResult, errorResult, formatMessages } from "../types.js";

export function register(server: McpServer, ctx: ToolContext) {
  // ── create_peer ─────────────────────────────────────────────────────
  server.registerTool(
    "create_peer",
    {
      description: [
        "Get or create a peer with the given ID.",
        "Use this to register a new participant (user or agent) in the workspace.",
        "Returns the peer ID and any configuration that was set.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("Unique identifier for the peer."),
        configuration: z
          .object({
            observeMe: z.boolean().nullable().optional().describe(
              "Whether derivation tasks should be created for this peer's messages. Default: true.",
            ),
          })
          .optional()
          .describe("Optional peer configuration."),
      },
    },
    async ({ peer_id, configuration }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id, { configuration });
        return textResult({ peer_id: peer.id, configuration: peer.configuration });
      } catch (e) {
        return errorResult(
          `Failed to create peer: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── list_peers ──────────────────────────────────────────────────────
  server.registerTool(
    "list_peers",
    {
      description: [
        "List all peers in the current workspace.",
        "Use this to discover which users and agents exist.",
        "Returns an array of peer IDs.",
      ].join("\n"),
      inputSchema: {},
    },
    async () => {
      try {
        const page = await ctx.honcho.peers();
        const peers: { id: string }[] = [];
        for await (const peer of page) {
          peers.push({ id: peer.id });
        }
        return textResult(peers);
      } catch (e) {
        return errorResult(
          `Failed to list peers: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── chat ────────────────────────────────────────────────────────────
  server.registerTool(
    "chat",
    {
      description: [
        "Ask a natural-language question about a peer's knowledge and get an answer from Honcho's reasoning system.",
        "Use this to query what Honcho knows about any peer — their preferences, history, personality, etc.",
        "Returns a natural-language answer, or 'None' if no relevant information exists.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The peer to query about."),
        query: z.string().describe("Natural-language question."),
        target_peer_id: z
          .string()
          .optional()
          .describe(
            "Optional: query what peer_id knows about this target peer instead of their global representation.",
          ),
        session_id: z
          .string()
          .optional()
          .describe("Optional: scope the query to a specific session."),
        reasoning_level: z
          .enum(["minimal", "low", "medium", "high", "max"])
          .optional()
          .describe("Reasoning effort. Higher = more detailed but slower."),
      },
    },
    async ({ peer_id, query, target_peer_id, session_id, reasoning_level }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const result = await peer.chat(query, {
          target: target_peer_id,
          session: session_id,
          reasoningLevel: reasoning_level,
        });
        return textResult(result ?? "None");
      } catch (e) {
        return errorResult(
          `Chat failed: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── get_peer_card ───────────────────────────────────────────────────
  server.registerTool(
    "get_peer_card",
    {
      description: [
        "Get the peer card — a compact set of biographical facts about a peer.",
        "Use this when you need a quick summary of who someone is.",
        "Returns an array of fact strings, or null if no card exists yet.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .optional()
          .describe(
            "Optional: get this peer's card about the target instead of their own.",
          ),
      },
    },
    async ({ peer_id, target_peer_id }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const card = await peer.getCard(target_peer_id);
        return textResult(card ?? "No peer card found.");
      } catch (e) {
        return errorResult(
          `Failed to get peer card: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── set_peer_card ───────────────────────────────────────────────────
  server.registerTool(
    "set_peer_card",
    {
      description: [
        "Set or update the peer card — a list of biographical facts about a peer.",
        "Use this to manually establish or correct facts about a peer.",
        "Returns the updated peer card.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The observer peer."),
        peer_card: z
          .array(z.string())
          .describe("Array of fact strings to set as the peer card."),
        target_peer_id: z
          .string()
          .optional()
          .describe(
            "Optional: set this peer's card about the target instead of their own.",
          ),
      },
    },
    async ({ peer_id, peer_card, target_peer_id }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const result = await peer.setCard(peer_card, target_peer_id);
        return textResult(result ?? "Peer card set successfully");
      } catch (e) {
        return errorResult(
          `Failed to set peer card: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── get_peer_context ────────────────────────────────────────────────
  server.registerTool(
    "get_peer_context",
    {
      description: [
        "Get comprehensive context for a peer — combines their representation (conclusions) and peer card.",
        "Use this when you need the full picture of what Honcho knows about someone.",
        "Returns an object with representation, peer_card, peer_id, and target_id.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .optional()
          .describe("Optional: get context about this target peer."),
        search_query: z
          .string()
          .optional()
          .describe("Optional: semantic search to filter relevant conclusions."),
        max_conclusions: z
          .number()
          .optional()
          .describe("Optional: max number of conclusions to include."),
      },
    },
    async ({ peer_id, target_peer_id, search_query, max_conclusions }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const context = await peer.context({
          target: target_peer_id,
          searchQuery: search_query,
          maxConclusions: max_conclusions,
        });
        return textResult({
          peer_id: context.peerId,
          target_id: context.targetId,
          representation: context.representation,
          peer_card: context.peerCard,
        });
      } catch (e) {
        return errorResult(
          `Failed to get peer context: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── get_representation ──────────────────────────────────────────────
  server.registerTool(
    "get_representation",
    {
      description: [
        "Get the formatted representation for a peer — a text summary built from their conclusions.",
        "Use this when you want the textual representation without the peer card.",
        "Returns a formatted string of conclusions.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The observer peer."),
        target_peer_id: z
          .string()
          .optional()
          .describe("Optional: get representation about this target peer."),
        session_id: z
          .string()
          .optional()
          .describe("Optional: scope to a specific session."),
        search_query: z
          .string()
          .optional()
          .describe("Optional: semantic search to filter conclusions."),
        max_conclusions: z
          .number()
          .optional()
          .describe("Optional: max number of conclusions."),
      },
    },
    async ({
      peer_id,
      target_peer_id,
      session_id,
      search_query,
      max_conclusions,
    }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const rep = await peer.representation({
          target: target_peer_id,
          session: session_id,
          searchQuery: search_query,
          maxConclusions: max_conclusions,
        });
        return textResult(rep);
      } catch (e) {
        return errorResult(
          `Failed to get representation: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── get_peer_metadata ───────────────────────────────────────────────
  server.registerTool(
    "get_peer_metadata",
    {
      description: [
        "Get the metadata dictionary for a peer.",
        "Use this to read custom attributes stored on a peer.",
        "Returns a JSON object of key-value pairs.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The peer to get metadata for."),
      },
    },
    async ({ peer_id }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const metadata = await peer.getMetadata();
        return textResult(metadata);
      } catch (e) {
        return errorResult(
          `Failed to get peer metadata: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── set_peer_metadata ───────────────────────────────────────────────
  server.registerTool(
    "set_peer_metadata",
    {
      description: [
        "Set metadata for a peer (overwrites existing metadata).",
        "Use this to store custom attributes on a peer.",
        "Returns a confirmation message.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The peer to set metadata for."),
        metadata: z
          .record(z.string(), z.unknown())
          .describe("Key-value pairs to set."),
      },
    },
    async ({ peer_id, metadata }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        await peer.setMetadata(metadata);
        return textResult("Peer metadata set successfully");
      } catch (e) {
        return errorResult(
          `Failed to set peer metadata: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );

  // ── search_peer_messages ────────────────────────────────────────────
  server.registerTool(
    "search_peer_messages",
    {
      description: [
        "Semantic search across all messages authored by a specific peer.",
        "Use this to find what a particular peer has said across all sessions.",
        "Returns an array of matching messages.",
      ].join("\n"),
      inputSchema: {
        peer_id: z.string().describe("The peer whose messages to search."),
        query: z.string().describe("Search query."),
      },
    },
    async ({ peer_id, query }) => {
      try {
        const peer = await ctx.honcho.peer(peer_id);
        const messages = await peer.search(query);
        return textResult(formatMessages(messages));
      } catch (e) {
        return errorResult(
          `Search failed: ${e instanceof Error ? e.message : String(e)}`,
        );
      }
    },
  );
}

File: mcp-rs/src/tools/peers.rs
use crate::config::HonchoConfig;
use crate::honcho_client::HonchoClient;
use crate::tools::{
    ToolError, enum_prop, number_prop, object_schema, optional_string, optional_u64, page_id_list,
    query_push_optional, required_array, required_string, resource_id, response_field_or_value,
    string_array_prop, string_prop,
};
use rmcp::model::JsonObject;
use serde_json::{Map, Value, json};

pub(crate) fn create_peer_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("Unique identifier for the peer")),
            (
                "configuration",
                json!({
                    "type": "object",
                    "properties": {
                        "observeMe": {
                            "type": ["boolean", "null"],
                            "description": "Whether derivation tasks should be created for this peer's messages."
                        }
                    },
                    "additionalProperties": true
                }),
            ),
        ],
        &["peer_id"],
    )
}

pub(crate) fn chat_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("The peer to query about")),
            ("query", string_prop("Natural-language question")),
            ("target_peer_id", string_prop("Optional target peer")),
            ("session_id", string_prop("Optional session scope")),
            (
                "reasoning_level",
                enum_prop(
                    "Reasoning effort",
                    &["minimal", "low", "medium", "high", "max"],
                ),
            ),
        ],
        &["peer_id", "query"],
    )
}

pub(crate) fn peer_target_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("The observer peer")),
            ("target_peer_id", string_prop("Optional target peer")),
        ],
        &["peer_id"],
    )
}

pub(crate) fn set_peer_card_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("The observer peer")),
            ("peer_card", string_array_prop("Array of fact strings")),
            ("target_peer_id", string_prop("Optional target peer")),
        ],
        &["peer_id", "peer_card"],
    )
}

pub(crate) fn get_peer_context_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("The observer peer")),
            ("target_peer_id", string_prop("Optional target peer")),
            (
                "search_query",
                string_prop("Optional semantic search query"),
            ),
            (
                "max_conclusions",
                number_prop("Optional max number of conclusions"),
            ),
        ],
        &["peer_id"],
    )
}

pub(crate) fn get_representation_schema() -> JsonObject {
    object_schema(
        vec![
            ("peer_id", string_prop("The observer peer")),
            ("target_peer_id", string_prop("Optional target peer")),
            ("session_id", string_prop("Optional session scope")),
            (
                "search_query",
                string_prop("Optional semantic search query"),
            ),
            (
                "max_conclusions",
                number_prop("Optional max number of conclusions"),
            ),
        ],
        &["peer_id"],
    )
}

pub(crate) async fn create_peer(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let mut body = json!({ "id": peer_id });
    if let Some(configuration) = args.get("configuration") {
        body["configuration"] = configuration.clone();
    }

    let peer: Value = client
        .post_json(&["workspaces", &config.workspace_id, "peers"], body)
        .await?;
    Ok(json!({
        "peer_id": resource_id(&peer),
        "configuration": peer.get("configuration").cloned().unwrap_or_else(|| json!({})),
    }))
}

pub(crate) async fn list_peers(
    client: &HonchoClient,
    config: &HonchoConfig,
) -> Result<Value, ToolError> {
    let page: Value = client
        .post_json(
            &["workspaces", &config.workspace_id, "peers", "list"],
            json!({}),
        )
        .await?;
    Ok(page_id_list(&page, "peers"))
}

pub(crate) async fn chat(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let query = required_string(args, "query")?;
    let mut body = json!({
        "query": query,
        "stream": false,
    });
    if let Some(target) = optional_string(args, "target_peer_id")? {
        body["target"] = json!(target);
    }
    if let Some(session_id) = optional_string(args, "session_id")? {
        body["session_id"] = json!(session_id);
    }
    if let Some(reasoning_level) = optional_string(args, "reasoning_level")? {
        body["reasoning_level"] = json!(reasoning_level);
    }

    let response: Value = client
        .post_json(
            &[
                "workspaces",
                &config.workspace_id,
                "peers",
                &peer_id,
                "chat",
            ],
            body,
        )
        .await?;
    Ok(response
        .get("content")
        .cloned()
        .filter(|value| !value.is_null())
        .unwrap_or_else(|| json!("None")))
}

pub(crate) async fn get_peer_card(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let mut query = Vec::new();
    query_push_optional(
        &mut query,
        "target",
        optional_string(args, "target_peer_id")?,
    );
    let response: Value = client
        .get_json(
            &[
                "workspaces",
                &config.workspace_id,
                "peers",
                &peer_id,
                "card",
            ],
            &query,
        )
        .await?;
    Ok(response
        .get("peer_card")
        .cloned()
        .filter(|value| !value.is_null())
        .unwrap_or_else(|| json!("No peer card found.")))
}

pub(crate) async fn set_peer_card(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let peer_card = Value::Array(required_array(args, "peer_card")?.clone());
    let mut query = Vec::new();
    query_push_optional(
        &mut query,
        "target",
        optional_string(args, "target_peer_id")?,
    );

    let response: Value = client
        .put_json_with_query(
            &[
                "workspaces",
                &config.workspace_id,
                "peers",
                &peer_id,
                "card",
            ],
            &query,
            json!({ "peer_card": peer_card }),
        )
        .await?;
    Ok(response
        .get("peer_card")
        .cloned()
        .unwrap_or_else(|| json!("Peer card set successfully")))
}

pub(crate) async fn get_peer_context(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let mut query = Vec::new();
    query_push_optional(
        &mut query,
        "target",
        optional_string(args, "target_peer_id")?,
    );
    query_push_optional(
        &mut query,
        "search_query",
        optional_string(args, "search_query")?,
    );
    if let Some(max_conclusions) = optional_u64(args, "max_conclusions")? {
        query.push(("max_conclusions", max_conclusions.to_string()));
    }

    let response: Value = client
        .get_json(
            &[
                "workspaces",
                &config.workspace_id,
                "peers",
                &peer_id,
                "context",
            ],
            &query,
        )
        .await?;
    Ok(json!({
        "peer_id": response.get("peer_id").cloned().unwrap_or_else(|| json!(peer_id)),
        "target_id": response.get("target_id").cloned().unwrap_or(Value::Null),
        "representation": response.get("representation").cloned().unwrap_or(Value::Null),
        "peer_card": response.get("peer_card").cloned().unwrap_or(Value::Null),
    }))
}

pub(crate) async fn get_representation(
    client: &HonchoClient,
    config: &HonchoConfig,
    args: &Map<String, Value>,
) -> Result<Value, ToolError> {
    let peer_id = required_string(args, "peer_id")?;
    let mut body = json!({});
    if let Some(target) = optional_string(args, "target_peer_id")? {
        body["target"] = json!(target);
    }
    if let Some(session_id) = optional_string(args, "session_id")? {
        body["session_id"] = json!(session_id);
    }
    if let Some(search_query) = optional_string(args, "search_query")? {
        body["search_query"] = json!(search_query);
    }
    if let Some(max_conclusions) = optional_u64(args, "max_conclusions")? {
        body["max_conclusions"] = json!(max_conclusions);
    }

    let response: Value = client
        .post_json(
            &[
                "workspaces",
                &config.workspace_id,
                "peers",
                &peer_id,
                "representation",
            ],
            body,
        )
        .await?;
    Ok(response_field_or_value(response, "representation"))
}
